Add Performance Tests (Fixes #19)

wallstop opened the pull request:
# Overview
Using the job system is great, and might very well be extremely performant. However, just using the job framework does not guarantee performance. There is the cost of marshaling from current state to buffers the job framework uses, then the cost of splitting out the work, then the cost of  marshaling the job framework data back to the text arrays.

I have several ideas about how to improve performance, but without benchmarks, you can't really say whether or not one idea is faster than another. Data doesn't lie 😎. The first step towards more perf is having data about the current state of the world.

# The Fix (Fixes #19)
I've done a few things here, all in service of having meaningful performance benchmarks.
&check; Migrated reflected fields to public/private access gated on whether or not you're in Unity Editor. This adds some safety and allows tests and editor scripts to modify state, which is needed. I understand that modifying these at runtime poses challenges (as you need to recompute buffers/state). Am open to alternatives, but this is a pattern that I've had success with in other projects.
&check; Added some simple performance tests that run modifiers across different inputs. I chose the input sequence of "A" repeated some number of times. "A" has an "interesting" number of vertices.
&check; I've changed all of the properties on existing CharModifier implementations to be public (with back-compat serialization attributes, so as not to break existing users). I don't see a reason to hide these, it can be useful to set these up programmatically (like I need to do in tests) and modifying the properties at runtime poses no issue, as all of the job system work is currently blocked on the main thread. There is no time in which the job framework does not have a consistent view of the various bits of state.
&check; Fixed several inconsistencies in null checks for texts in TweenManager - some loops were doing this, some were not
&check; Fixed a bug where removing the last text instance from a TweenManager would throw an exception
&check; Fixed a bug where empty text would throw
&check; Moved (most) usings of changed files into inside namespace - this has the highest chance of not breaking users. When usings are outside of the namespace, depending on how things are named, this has the potential to break client code.
&check; Added current benchmarks to README. The benchmark tests are structured in such a way that you can just copy-paste the output from Unity's Test Runner into the README and it'll be a valid markdown table.

# Ideas for Perf Improvement
Note: All of these need testing

1. See how fast a simple for loop is. It might be faster to just for loop over stuff if the size of the data is less than some threshold.
2. Change batch size from 64 to some other number (dynamically?)
3. Instead of generating temp arrays every call, cache persistent arrays and just memcpy data to them

# PR Feedback
If the above changes are too much to review/take all at once, but you're interested in taking some/all of them, I can break this PR up into smaller ones and stage them however you see fit, just let me know!
Also, I'm running CSharpier with default settings, if you have settings like a different line length or whatever, I can try to figure out how to set things up such that the config is checked into the repo and available for all contributors.

wallstop commented on Runtime/TweenManager.cs:
```diff
@@ -211,6 +244,21 @@ private void CreateCharDataArray()
         private void ApplyModifiers(float progress)
         {
+            bool anyTexts = false;
+            foreach (TMP_Text text in _texts)
```
Please note: foreach loops are pretty optimized, clarify intent, and are hard to get wrong. They are also generally faster, as the typical thing I'm seeing in this code base is re-indexing into arrays/lists - this is more expensive than just grabbing a reference to an element and using it.

https://gist.github.com/wallstop/df1ffcac37641c03870d88bed7fdeae2

| Method           | Mean     | Error     | StdDev    |
|----------------- |---------:|----------:|----------:|
| ForLoopArray     | 4.034 ms | 0.0787 ms | 0.0842 ms |
| ForEachLoopArray | 4.202 ms | 0.0691 ms | 0.0647 ms |
| ForLoopList      | 4.170 ms | 0.0797 ms | 0.1007 ms |
| ForEachLoopList  | 3.893 ms | 0.0500 ms | 0.0467 ms |

This data shows that a foreach loop on a list is actually *faster* than a counting for loop.

Please note, the foreach loop does *not* generate any garbage - it calls .GetEnumerator() on the collection implicitly, which for almost all standard lib collections, is a value type, not a reference type, and is stack allocated.

AlicanHasirci replied:
Guess I'll be swinging for List+ForEach from now on 👍

wallstop commented on Runtime/Modifiers/TransformModifier.cs:
```diff
@@ -1,45 +1,46 @@
-            Rotation,
-            Scale,
-        }
+        Position = 0,
```
Added explicit values such that, if things are ever reordered or changed, it does not break existing users.

AlicanHasirci replied:
Don't think these will ever change but none the less, It is a good practice to add ordinals 👍 
Maybe we can rename this to TransformType would be more descriptive

wallstop commented on Runtime/Modifiers/TransformModifier.cs:
```diff
@@ -1,45 +1,46 @@
-            Z = 1 << 2,
-        }
+    [Flags]
+    public enum ModifierScale
```
Can change these names to be whatever you want

AlicanHasirci replied:
'ScaleMask' makes more sense in here but we can actually get rid of this enum and property by making the scale transactional like rotation and position.

wallstop replied:
I think I've captured your intent here, but I don't quite know what `transactional` means in this context 😅

AlicanHasirci replied:
My bad, transactional is bit vague.
The implementation for `position` and `rotation` vs `scale` is a bit different if you look into it, hence the requirement for a scale mask. 
For example, lets we have a `TransformModifier` with `Intensity = float3(0, 1, 0)`. When progress is set to 1:
* **Position**: original `transform.position.y` gets increased by 1
* **Rotation**: original `transform.eulerAngles.y` gets increased by 1
* **Scale**: `transform.localeScale` gets **set** to (0,1,0)
You see, when `TransformType` is `position` or `rotation` the intensity value is considered as the _amount changed_; when it is scale, the intensity is the _final value_ of scale so we need a _mask_ to avoid changing certain dimensions of scale.
This issue is out of this PR's scope though 😌

wallstop commented on Runtime/TweenManager.cs:
```diff
@@ -249,8 +299,14 @@ NativeArray<float4> colors
+                    return;
+                }
+
                 text.mesh.SetVertices(vertices, offset, count);
```
This thing blows up if you call it with count = 0, from experimentation

wallstop commented:
I had found a bug that I thought I introduced, but it appears to be latent. Pre-emptively closed this PR thinking the bug was in this code, but have confirmed that it's in main. Will open an issue and try to fix.

AlicanHasirci commented:
Benchmarks were definitely in my roadmap since if ends does not justify the means this repo suddenly becomes redundant 😄 So you've saved me from a lot of trouble.
No need to break it down, I am looking into it.

AlicanHasirci commented on Editor/TextTweenEditor.cs:
```diff
@@ -4,7 +4,6 @@ namespace TextTween.Editor
     using System;
```
residual using directive

AlicanHasirci commented on Runtime/TweenManager.cs:
```diff
@@ -1,16 +1,16 @@
-using System;
-using System.Collections.Generic;
-using System.Runtime.CompilerServices;
-using TextTween.Native;
-using TMPro;
-using Unity.Collections;
-using Unity.Jobs;
-using Unity.Mathematics;
-using UnityEngine;
-using Object = UnityEngine.Object;
```
I've seen by you before, adding all the using directives inside namespace, is there a specific reason?

AlicanHasirci replied:
That is a great advice! Accidentally creating an "Editor" namespace is so easy and happened to me a few times as well.

AlicanHasirci commented on Runtime/TweenManager.cs:
```diff
@@ -21,11 +21,25 @@ public class TweenManager : MonoBehaviour, IDisposable
+         */
         [SerializeField]
-        private TMP_Text[] _texts;
+#if UNITY_EDITOR
+        public
+#else
+        private
+#endif
+        TMP_Text[] _texts = Array.Empty<TMP_Text>();
```
We can actually achieve this by making these fields internal and making use of `[assembly: System.Runtime.CompilerServices.InternalsVisibleTo("")]` this can reduce the cognitive load here and provide a better encapsulation.

wallstop replied:
Love the approach, will use in future projects. For some reason I thought this didn't work with Unity assembly definitions, but it does! That's rad.

wallstop commented:
Will iterate on this PR once the text change / enable/disable fixes are in, I think there's some overlap and code removal I can do here if those problems are handled.

AlicanHasirci commented:
LGTM. I think we can merge this after resolving the conflicts 👍

wallstop commented:
> Benchmarks were definitely in my roadmap since if ends does not justify the means this repo suddenly becomes redundant 😄 So you've saved me from a lot of trouble. No need to break it down, I am looking into it.

Updates since last review:
1. I've rebased main to get the latest and greatest changes
2. I've stripped out all of the unnecessary code that I could find
3. Removed weird preprocessor directive private/public scope, replaced with internal + internals visible to
4. Updated custom Editor to take advantage of newly fields
5. Updated performance tests to take advantage of the newly accessible
6. Fixed a bug in the recent disabled/enabled PR - if a TweenManager started out disabled, it would toggle on correctly, but then would stay on when toggled off, until you toggled it on then off again. This was due to me not properly setting up initial state when the object is disabled
7. Fixed a bug in the Editor to check for null TweenManager, which appears to happen somewhere around the time you switch from Editor mode to running the Test in Test Runner, but are still inspecting a TweenManager
8. Fixed a bug in "Find All Texts / Find All Modifiers". For some reason, on Windows 11, Unity 2022.3.51f1, the "target" can be null (and the cast to TweenManager was throwing, since it's null). To remedy this, instead of relying on target, rely on the serialized property. We can just look at the `serializedObject` its a part of and derive our target. 
9. Changed the property name detection of custom editor context mode to use nameof, now that we have access to the properties. The benefit of nameof is that it's linked into language controls. Without nameof, if the literal name of the variable `_texts` is changed, this code would break *silently*. With nameof, it's a reference to the variable that is resolved to a constant at compile time. So you can refactor/rename all you want, either the `nameof(TweenManager._texts)` will update, or, your code won't compile and you'll have to fix it. Either way, no bugs.
10. Added some minor handling to improve the case where, if users had multiple texts, but some text was empty, certain copy operations would fail due to the amount being copied being 0. Previous code seems to be built towards the assumption that there was only 1 text, which already has similar short-circuits built in.

Please let me know if you would like any of the above split out or in different PRs, and/or disagree with any of the changes made. Some of them were necessary to get switching back and forth from a Test Runner context working without something breaking, but definitely not all of them.

wallstop commented on Editor/TweenManagerEditor.cs:
```diff
@@ -187,7 +193,7 @@ private void FindMissingComponents<T>(SerializedProperty serializedProperty)
             }
 
-            serializedObject.ApplyModifiedProperties();
+            serializedProperty.serializedObject.ApplyModifiedProperties();
```
This thing was also null when the target was null, so had to use the pointer from the property

AlicanHasirci commented:
> > Benchmarks were definitely in my roadmap since if ends does not justify the means this repo suddenly becomes redundant 😄 So you've saved me from a lot of trouble. No need to break it down, I am looking into it.
> 
> Updates since last review:
> 
> 1. I've rebased main to get the latest and greatest changes
> 2. I've stripped out all of the unnecessary code added in this PR that I could find
> 3. Removed odd preprocessor directive private/public scope, replaced with internal + internals visible to
> 4. Updated custom TweenManager Editor to take advantage of newly accessible fields
> 5. Updated performance tests to take advantage of newly accessible fields
> 6. Fixed a bug in the recent disabled/enabled PR - if a TweenManager started out disabled, it would toggle on correctly, but then would stay on when toggled off, until you toggled it on then off again. This was due to me not properly setting up initial state when the object is disabled
> 7. Fixed a bug in the Editor to check for null TweenManager, which appears to happen somewhere around the time you switch from Editor mode to running the Test in Test Runner, but are still inspecting a TweenManager
> 8. Fixed a bug in "Find All Texts / Find All Modifiers". For some reason, on Windows 11, Unity 2022.3.51f1, the "target" can be null (and the cast to TweenManager was throwing, since it's null). To remedy this, instead of relying on target, rely on the serialized property. We can just look at the `serializedObject` its a part of and derive our target.
> 9. Changed the property name detection of TweenManager custom editor context mode to use nameof, now that we have access to the properties. The benefit of nameof is that it's linked into language controls. Without nameof, if the literal name of the variable `_texts` is changed, this code would break _silently_. With nameof, it's a reference to the variable that is resolved to a constant at compile time. So you can refactor/rename all you want, either the `nameof(TweenManager._texts)` will update (assuming you have a modern IDE), or, it won't update and your code won't compile. Either way, no bugs.
> 10. Added some minor handling to improve the case where, if users had multiple texts in a TweenManager, but some text was empty, certain copy operations would fail due to the amount being copied being 0. Previous code seems to be built towards the assumption that there was only 1 text, which already has similar short-circuits built in.
> 11. I've applied enum renames
> 12. Added comments around anything that didn't seem obvious, but happy to add more if you think there are any bits that could use it.
> 
> Please let me know if you would like any of the above split out or in different PRs, and/or disagree with any of the changes made. Some of them were necessary to get switching back and forth from a Test Runner context working without something breaking, but definitely not all of them.

Thanks a lot for thorough explanation, these all look good to me, with some. great suggestions and catches. The only thing suspect to change would be the names of the internal variables. Since they are not privates anymore we can remove the underscores, but it is a minor issue not worth to make this PR wait anymore so I'll merge this and change those later.
